clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -256,11 +256,13 @@ public void poll(long timeout, long now, PollCondition pollCondition) {
      * nor will it execute any delayed tasks.
      */
     public void pollNoWakeup() {
-        disableWakeups();
-        try {
-            poll(0, time.milliseconds(), null);
-        } finally {
-            enableWakeups();
+        synchronized (this) {
```

Comment:
@ijuma tried it out, practically (1000s of iterations :D) and logically this looks fine too:

```java
    /**
     * Poll for network IO and return immediately. This will not trigger wakeups,
     * nor will it execute any delayed tasks.
     */
    public void pollNoWakeup() {
        synchronized (this) {
            synchronized (this.wakeup) {
                disableWakeups();
                try {
                    poll(0, time.milliseconds(), null);
                } finally {
                    enableWakeups();
                }
            }
        }
    }
```

combined with simply:

```java
    public void maybeTriggerWakeup() {
        synchronized (this.wakeup) {
            if (wakeupDisabledCount == 0 && wakeup.get()) {
                log.trace("Raising wakeup exception in response to user wakeup");
                wakeup.set(false);
                throw new WakeupException();
            }
        }
    }
```

This should rule out potential deadlocking from `fireCompletedPendingRequests` right?